Focus control upstream

meguiraun opened the pull request:
hi guys,

This is an attempt to bring Biomax's old focus control element into latest mxcube. 

WIP because your input is needed on how we handle these kind of components, it technically works as it is. 

- Added a new BeamFocus component, it currently has Biomax specific logic but i did not want to generalise too much too early
- Configurable "beam control" elements in the ui

<img width="132" alt="Screenshot 2024-03-04 at 08 48 34" src="https://github.com/mxcube/mxcubeweb/assets/12758327/7c42261d-b329-4113-acf7-0481755ce79d">

meguiraun committed:
adding beam focus component

meguiraun committed:
configurable beam control (phase, aperture, focus)

meguiraun committed:
missing props

meguiraun committed:
handle changes

meguiraun committed:
fix argument passing

meguiraun committed:
fix beamfocus url

meguiraun committed:
ignore motor states

meguiraun committed:
adding pnpm-lock.yaml

meguiraun committed:
no console logs

axelboc commented on ui/package.json:
```diff
@@ -59,6 +59,7 @@
     "react-router": "^6.2.1",
     "react-router-bootstrap": "^0.26.1",
     "react-router-dom": "^6.19.0",
+    "react-select": "5.8.0",
```
Hey @meguiraun, you should use Bootstrap's `Form.Select` component instead. This is what is used for other drop-down selectors. For instance, check out the code for the "Beam size" selector (the component is called `ApertureInput`).

meguiraun replied:
mmm.... I cannot customize the style of the different options. 

Based on the currently selected beam focus, the options for the aperture will be displayed with a different color or even disabled.

react-select allows me this customization, but I am struggling with Form.Select
<img width="781" alt="image" src="https://github.com/mxcube/mxcubeweb/assets/12758327/47513ea2-1e66-475b-a6a9-3f96bfd3f4fa">

axelboc replied:
Indeed, in that case you should use Bootstrap's `Dropdown` component. https://react-bootstrap.netlify.app/docs/components/dropdowns#customization

meguiraun replied:
you are my hero 🤗

meguiraun committed:
better bootstrap element

meguiraun committed:
remove react-select

marcus-oscarsson commented:
Just a question, how is this control used, is to de-focus the beam to change the beam size ?

meguiraun commented:
> Just a question, how is this control used, is to de-focus the beam to change the beam size ?

The user selects a beam size appropriate for their crystal size, which corresponds to different beam (de)focus

meguiraun committed:
adding beam definer to beam info data

meguiraun committed:
generalize beam definer component

meguiraun committed:
beam definer component

meguiraun commented:
I refactored few things here and there. In essence I added the beam definer information into the beamInfo info dict, and propagate it to the UI.

I based on the ID232 beam definer approach as NState hwobj, and adapted from my previous attempt, now "100x100" is one option in the definer object.

One thing that I have not solved is how to generalize the UI styling (see [this](https://github.com/mxcube/mxcubeweb/blob/focus_control_upstream/ui/src/components/SampleView/BeamDefinerInput.js#L39) ), any suggestion is welcome

I also created a new beam definer mockup:
https://github.com/mxcube/mxcubecore/pull/872

![beam_definer](https://github.com/mxcube/mxcubeweb/assets/12758327/c54192b8-6fde-4907-b919-fac2a3ef7b0c)

meguiraun committed:
linting does not hurt

meguiraun committed:
dont mess around with existing components

meguiraun committed:
custom styling

pass to the ui custom styling if available in the hwobj

meguiraun committed:
added beam definer mockup to beam mockup

meguiraun commented:
I moved down into the hwobj the custom css styling.

please, have a look!

axelboc commented on ui/src/components/SampleView/BeamDefinerInput.js:
```diff
@@ -0,0 +1,126 @@
+      col = 'black';
+    }
+    return col;
+  }
```
Nothing critical, but I can think of a few ways to refactor this:

```js
  itemColor(status) {
    if (status === 'disable') {
      return 'red';
    }

    if (status === 'warning') {
      return '#fc9003'; // darker orange
    }

    return 'black';
  }
```

... or with a record:

```js
// at the top of the file
const ITEM_COLOR = {
  disable: 'red',
  warning: '#fc9003',
}

// in the `JSX` method:
fontColor: ITEM_COLOR[status] || 'black';
```

... or even better, to keep JS and CSS neatly separated, you can just store the `status` inside each `option` object and then set a `data-` attribute on each `Dropdown.Item`:

```jsx
import styles from './BeamDefinedInput.module.css'; // CSS Module to avoid global styles; otherwise put the styles in `SampleView.css` and just pick a very unique class name for the options

<Dropdown.Item
  // ...
  className={styles.option} // 
  data-status={status}
/>
```

 Then, in the CSS:

```css
.option {
  /* color: black; */ /* probably the default color, so can be left out */
  float: left; /* you have the `float-left` class in the JSX, which seems strange to me, but there's probably a good reason I'm not seeing */
}

.option[data-status="disable"] { color: red; }
.option[data-status="warning"] { color: #fc9003; }
```

meguiraun replied:
I simplify this logic from your suggestion. I could not make it work with css though

axelboc commented on ui/src/components/SampleView/BeamDefinerInput.js:
```diff
@@ -0,0 +1,126 @@
+import './SampleView.css';
+import cx from 'classnames';
+
+export default class BeamDefinerInput extends React.Component {
```
This should be quite straightforward to write as a functional component if you're up for it.

meguiraun replied:
I am up for a challengue! good suggestion!

meguiraun replied:
functionalized!

axelboc commented on ui/src/containers/SampleViewContainer.js:
```diff
@@ -126,24 +137,44 @@ class SampleViewContainer extends Component {
+                <BeamDefinerInput
+                  beamDefinerInputList={this.props.sampleViewState.definerList}
+                  currentDefiner={this.props.sampleViewState.currentDefiner}
+                  customStyling={this.props.sampleViewState.customStyling}
```
I feel like the logic in the `render` method of `BeamDefinerInput` would be much simpler if you passed this through the `apertureList` prop:

```jsx
apertureList={this.props.sampleViewState.customStyling || this.props.sampleViewState.apertureList}
```

This way, you can have a single loop to create the `optionList` array (you may not even need this `optionList` array at all, since the loop below can be moved directly into the JSX):

```jsx
const optionList = Object.entries(this.props.apertureList).map((key, value) => {
  // key is the label (e.g. "50x50")
  // value is either the same as the label or an object when `customStyling` is defined, so you can do:
  if (typeof value === 'object') {
    return { ... }; // or return a `<Dropdown.Item />` directly if you move the loop inside the JSX
  }
  return { ... }
});
```

axelboc commented on ui/src/containers/SampleViewContainer.js:
```diff
@@ -126,24 +137,44 @@ class SampleViewContainer extends Component {
-                  state={diffractometerHo.state}
-                />
-              </div>
+              {phase.show ? (
```
The `phase.show && (` syntax can save you from using a ternary.

axelboc reviewed (commented):
Looking great! I've got lots of suggestions but don't feel like you have to resolve them — it's all coding style, nothing major.

meguiraun committed:
make if functional; simplify logic a bit

meguiraun committed:
simplify ternary using &&

meguiraun committed:
not needed as it is part of the beam object

meguiraun committed:
more robust when beam definer is not defined (pun)

meguiraun committed:
blacking

meguiraun committed:
extra check

marcus-oscarsson commented:
Just a status check, where are we with this one @meguiraun ?

fabcor-maxiv commented:
@marcus-oscarsson @meguiraun Is it okay to delete the branch?
